package.json

```diff
@@ -67,9 +63,6 @@
     ]
   },
   "browserslist": [
-    ">0.2%",
-    "not dead",
-    "not op_mini all",
-    "ie 11"
+    "node 6"
```

Comment:
We don't just support node 6 though. We also support these browsers. For example, `@testing-library/dom`  uses `aria-query` and `@testing-library/dom` needs to be able to run in a browser.

IE11 might be debateable but if we can convert spread syntax we might as well convert it, instead of having to release a new major.

Reply:
I don't see `@testing-library/dom` listing as supporting IE11, so I assume it only runs in modern browsers? (`@testing-library/dom` doesn't list any browsers it supports at all, no matter how well I searched for it)

This package also doesn't explicitly list which browsers it supports. What do you think of updating this package to officially support Node 6 and IE11? (I can either update the README in this PR to state the official browser support or create a follow-up PR, whatever you prefer) Hopefully IE11 support doesn't incur as heavy of a cost as supporting ancient browsers that are included in the current list, while still being maximally compatible with users.

Reply:
The question is what we gain by dropping support. I don't think we need to drop support for the listed browsers but rather rework the implementation.

Reply:
I think the biggest benefit is a significantly reduced network usage of the runtime dependency. I did some quick calculations and not depending on the runtime would reduce this package direct network impact by ~3000 Terabytes per year (8,5M downloads per week * 7MB for the runtime dependency * 52), as well as improved runtime performance (running polyfilled code is usually significantly slower than the native equivalents).

To show the difference, this is the before compiled output (here you can see that basically any language construct that is used is transformed, rather than only those that are necessary for the `browserslist`):

<details>

```js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _set = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/set"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/slicedToArray"));

var _entries = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/entries"));

var _find = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/find"));

var _stringify = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/json/stringify"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/keys"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/toConsumableArray"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/map"));

var _rolesMap = _interopRequireDefault(require("./rolesMap"));

var _context;

var elementRoleMap = new _map.default([]);
(0, _forEach.default)(_context = (0, _toConsumableArray2.default)((0, _keys.default)(_rolesMap.default).call(_rolesMap.default))).call(_context, function (key) {
  var role = _rolesMap.default.get(key);

  if (role) {
    var _context2, _context3;

    (0, _forEach.default)(_context2 = (0, _concat.default)(_context3 = []).call(_context3, (0, _toConsumableArray2.default)(role.baseConcepts), (0, _toConsumableArray2.default)(role.relatedConcepts))).call(_context2, function (relation) {
      if (relation.module === 'HTML') {
        var concept = relation.concept;

        if (concept) {
          var _context4;

          var conceptStr = (0, _stringify.default)(concept);
          var roles = ((0, _find.default)(_context4 = (0, _toConsumableArray2.default)((0, _entries.default)(elementRoleMap).call(elementRoleMap))).call(_context4, // eslint-disable-next-line no-unused-vars
          function (_ref) {
            var _ref2 = (0, _slicedToArray2.default)(_ref, 2),
                key = _ref2[0],
                value = _ref2[1];

            return (0, _stringify.default)(key) === conceptStr;
          }) || [])[1];

          if (!roles) {
            roles = new _set.default([]);
          }

          roles.add(key);
          elementRoleMap.set(concept, roles);
        }
      }
    });
  }
});
var _default = elementRoleMap;
exports.default = _default;
```

</details>

And this would be the new compiled output (with IE11 support dropped):

<details>

```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _rolesMap = _interopRequireDefault(require("./rolesMap"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const roleElementMap = new Map([]);
[..._rolesMap.default.keys()].forEach(key => {
  const role = _rolesMap.default.get(key);

  if (role) {
    [...role.baseConcepts, ...role.relatedConcepts].forEach(relation => {
      if (relation.module === 'HTML') {
        const concept = relation.concept;

        if (concept) {
          const relationConcepts = roleElementMap.get(key) || new Set([]);
          relationConcepts.add(concept);
          roleElementMap.set(key, relationConcepts);
        }
      }
    });
  }
});
var _default = roleElementMap;
exports.default = _default;
```

</details>

If you would retain IE11 support, this would be the output:

<details>

```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _rolesMap = _interopRequireDefault(require("./rolesMap"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var roleElementMap = new Map([]);

_toConsumableArray(_rolesMap.default.keys()).forEach(function (key) {
  var role = _rolesMap.default.get(key);

  if (role) {
    [].concat(_toConsumableArray(role.baseConcepts), _toConsumableArray(role.relatedConcepts)).forEach(function (relation) {
      if (relation.module === 'HTML') {
        var concept = relation.concept;

        if (concept) {
          var relationConcepts = roleElementMap.get(key) || new Set([]);
          relationConcepts.add(concept);
          roleElementMap.set(key, relationConcepts);
        }
      }
    });
  }
});

var _default = roleElementMap;
exports.default = _default;
```

</details>

In terms of file sizes (measured with `perl -le 'map { $sum += -s } @ARGV; print $sum' -- lib/`):

- Only Node 6 support 116Kb
- Current support with all ancient browsers 186Kb (but you would have to account for the 7MB dependency as well)
- Node 6 + IE11 support 176Kb, but without the runtime dependency (about 50% bigger than only Node 6)

That is all to say, you will sadly not be able to rework the implementation to get rid of all compiled output. Since this package transpiles from ES modules to CommonJS, some transformations will always be required. That said, if you remove the runtime, you actually end up with a smaller package (10Kb reduction) and you would remove the 7MB dependency.

Therefore, as a middleground, what do you think about removing the runtime, but leaving the `browserslist` as it is today? That would still save us the ~3000 Terabytes of network traffic, while we maintain compatibility with all relevant browser/runtimes and don't break users.

Reply:
Oh, with `Node 6 + IE 11` I actually mean the existing `browserslist`. E.g. transpiling with the current `browserslist` produces equivalent output as if you would only list `Node 6 + IE 11`.

Reply:
> The concern about @testing-library/dom is valid, but they use jsdom which is not a browser (it just provides browser-like APIs to nodejs code), so it retains all the compatibility with the node version.

This is incomplete. `@testing-library/dom` doesn't use `jsdom`. We support all sorts of environments. From browsers to `jsdom`.